ShapeInput from subset of atoms in molecule

DavidACosgrove opened the pull request:
#### Reference Issue
No issue.

#### What does this implement/fix? Explain your changes.
This allows the ShapeInput object optionally to be made from a subset of atoms in a molecule.
It also optionally allows the specification of a radius for dummy atoms and their inclusion in the ShapeInput.

#### Any other comments?
Both of these will be used in the shape searching of synthon space.

greglandrum reviewed (changes requested):
Rather than starting down the trail of adding a bunch of optional arguments to `PrepareConformer()`, I would prefer that we immediately move to introducing a struct with the options and then change the signature to `ShapeInput PrepareConformer(const RDKit::ROMol &mol, const ShapeInputOptions &opts=ShapeInputOptions())`

DavidACosgrove commented:
The 2 overloads of AlignMolecule, which also call PrepareConformer will also need to take the ShapeInputOptions, I assume.  Shall I roll the other options (useColors, max_preiters etc.) into the struct as well, and do we need to preserve the existing interfaces for backwards compatibility?

DavidACosgrove commented:
I've added the ShapeInputOptions class, added @nmaeder's attribute protection thing (very cool, I've been wanting something like that for ages!) and an extra option for using non-standard radii for atoms.  Also added a function for overlaying a ShapeInput onto a ShapeInput because it will be needed in the Synthon Space searching and a function for taking the transformation for same and applying it to a conformer.

DavidACosgrove commented:
As currently written, if, for example, the subset incorporates half a benzene ring, the color feature is still created, centred on the average coords of the selected atoms.  Is that sensible, or should the whole feature be skipped?  I am in 2 minds but erring towards the latter, I think.  This is relevant to the shape searching in synthon space where it is a likely occurrence.

DavidACosgrove commented:
So I've answered my own question above - partial color features won't be allowed.  Also added an option to exclude some atoms from color features.

greglandrum reviewed (changes requested):
Minor changes suggested

DavidACosgrove commented on External/pubchem_shape/PubChemShape.cpp:
```diff
@@ -246,82 +244,172 @@ ShapeInput PrepareConformer(const ROMol &mol, int confId, bool useColors) {
+}
+bool atomAllowedInColor(unsigned int atomIdx,
+                        const ShapeInputOptions &shapeOpts) {
+  return std::ranges::find(shapeOpts.notColorAtoms, atomIdx) ==
```
ChatGPT assured me that using std::ranges is better, but I can't now remember why.  I guess there must be a good reason for what is essentially duplicated functionality with a different API.

DavidACosgrove commented on External/pubchem_shape/PubChemShape.cpp:
```diff
@@ -246,82 +244,172 @@ ShapeInput PrepareConformer(const ROMol &mol, int confId, bool useColors) {
-    unsigned int Z = mol.getAtomWithIdx(i)->getAtomicNum();
-    if (Z > 1) {
+double getAtomRadius(unsigned int atomIdx, const ShapeInputOptions &shapeOpts) {
+  auto it = std::ranges::find_if(
+      shapeOpts.atomRadii,
+      [atomIdx](const auto &p) -> bool { return p.first == atomIdx; });
+  if (it == shapeOpts.atomRadii.end()) {
+    return -1.0;
+  }
+  return it->second;
+}
```
I normally try to sneak one through when you're not looking.  Disappointed that I missed the opportunity.

DavidACosgrove commented on Code/RDBoost/Wrap.h:
Sorry.  My IDE runs clang-format when I save a file.  I guess that wasn't done before that PR was accepted.

DavidACosgrove commented:
@greglandrum Merging in master required some additional changes.

DavidACosgrove commented:
The failure isn't down to this PR.  It failed to fetch boost as far as I can make out.

greglandrum reviewed (approved):
LGTM
